src/Carcassonne/Game/Game.cpp

```diff
@@ -518,4 +523,249 @@ std::pair<Direction, int> Game::move_score(Player player, TileType tile_type, Ti
+     case Direction::WestNorth:
+       *std::next(output_it, 8) = true;
+       role = 0;
+       break;
+   }
+   advance(output_it, 9);
+
+   // figure role: neurons_to_set = 4
+   *std::next(output_it, role) = true;
+   return result;
+}
+
+
+
+mb::result<mb::empty> Game::board_to_caffe_X(std::vector<bool>& output) const noexcept {
+   // tiles = g_board_width * g_board_height * sum_of:
+      // edges          = 4 * city_or_field_or_path[3]
+      // field edges    = 8 * city_or_field[3]
+      // contacts       = g_contact_types_count[24]
+      // connections    = g_connection_types_count[22]
+      // pennant        = true_or_false[1]
+      // moastery       = true_or_false[1]
+      // figure         = npcs[2]
+      // figure posit.  = pos[9]
+      // figure role    = at_city_or_field_or_path_or_moastery[4]
+   // remaining figures          = figs[7] * npcs[2]
+   // current tile               = sum_of_above[87]
+   // remaining tiles            = m_tile_set[70]
+   // best score <1<2<3...<255   = g_max_possible_score[255]
+   // RLPlayer score <1<2...<255 = g_max_possible_score[255] * npcs[2]
+   auto result = mb::ok;
+   const size_t tile_features_count = (4*3) + (8*3) + 24 + 22 + 1 + 1 + m_player_count + 9 + 4; // 87;
+   const size_t board_features_count = (g_board_width * g_board_height * tile_features_count)
+       + (g_initial_figures_count * m_player_count) + tile_features_count + m_tile_set.size()
+       + 2 * g_max_possible_score; // 166'784;
+   if (output.size() == 0) {
+      output = std::vector<bool>(board_features_count, false);
+   } else {
+      std::fill(output.begin(), output.end(), false);
+   }
+   auto output_it = output.begin();
+   for (int x = 0; x < g_board_width; ++x) {
+      for (int y = 0; y < g_board_height; ++y) {
+         const TilePlacement& placed_tile = board().tile_at(x,y);
+         if (0 == placed_tile.type) continue;
+         int neurons_offset = (x + y * g_board_width) * tile_features_count;
+         output_it = std::next(output.begin(), neurons_offset);
+         if (auto res = tile_to_caffe(placed_tile.tile(), output_it); !res.ok()) {
+            fmt::print("[tile_to_caffe internal error]: {}\n", res.msg());
+            return res;
+         }
+         auto output_edges_it = std::next(output.begin(), neurons_offset);
+         output_it = std::next(output.begin(), neurons_offset + tile_features_count - (m_player_count + 9 + 4));
+         if (auto res = figures_to_caffe(m_figures, x, y, output_edges_it, m_player_count, output_it); !res.ok()) {
+            fmt::print("[figures_to_caffe internal error]: {}\n", res.msg());
+            return res;
+         }
+      }
+   }
+   output_it = std::next(output.begin(), (g_board_height * g_board_width) * tile_features_count);
+   
+   // remaining figures: neurons_to_set = m_player_count * g_initial_figures_count
+   for (int i = 0; i < m_player_count; ++i) {
+      *std::next(output_it, m_figure_count[i]) = true;
```

Comment:
trochę mało czytelne te operacje na iteratorach

Reply:
W tym przypadku to jest to tylko aktywacja neuronu odpowiadającego ilości figur, które zostały graczowi 0, przesuwanie o ilość wszystkich figur i aktywacja neuronu gracza 1, znowu przesuwanie, itd. Możesz zaproponować coś czytelniejszego.
PS. Można by kiedyś pomyśleć o uniezależnieniu architektury sieci od liczby graczy.